kernel/src/sched/sched_class/mod.rs

```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: MPL-2.0
+
+#![warn(unused)]
+
+use alloc::{
+    boxed::Box,
+    sync::{Arc, Weak},
+};
+use core::fmt;
+
+use ostd::{
+    cpu::{num_cpus, CpuSet, PinCurrentCpu},
+    sync::{PreemptDisabled, SpinLock, SpinLockGuard},
+    task::{
+        scheduler::{inject_scheduler, EnqueueFlags, LocalRunQueue, Scheduler, UpdateFlags},
+        Task,
+    },
+    trap::disable_local,
+};
+
+mod time;
+
+mod fair;
+mod idle;
+mod real_time;
+mod stop;
+
+use ostd::arch::read_tsc as sched_clock;
+
+use super::priority::Priority;
+use crate::thread::Thread;
+
+#[allow(unused)]
+pub fn init() {
+    inject_scheduler(Box::leak(Box::new(ClassScheduler::default())));
+}
+
+/// Represents the middle layer between scheduling classes and generic scheduler
+/// traits. It consists of all the sets of run queues for CPU cores. Other global
+/// information may also be stored here.
+pub struct ClassScheduler {
+    rqs: Box<[SpinLock<PerCpuClassRqSet>]>,
+}
+
+/// Represents the run queue for each CPU core. It stores a list of run queues for
+/// scheduling classes in its corresponding CPU core. The current task of this CPU
+/// core is also stored in this structure.
+struct PerCpuClassRqSet {
+    stop: Arc<stop::StopClassRq>,
+    real_time: real_time::RealTimeClassRq,
+    fair: fair::FairClassRq,
+    idle: idle::IdleClassRq,
+    current: Option<Arc<Task>>,
+}
+
+/// The run queue for scheduling classes (the main trait). Scheduling classes
+/// should implement this trait to function as expected.
+trait SchedClassRq: Send + fmt::Debug {
+    type Entity;
+
+    /// Enqueues a task into the run queue.
+    fn enqueue(
+        &mut self,
+        thread: Arc<Thread>,
+        entity: SpinLockGuard<'_, SchedEntity, PreemptDisabled>,
+    );
+
+    /// Dequeues a task from the run queue.
+    fn dequeue(&mut self, entity: &Self::Entity);
```

Comment:
I found the interface of `SchedClassRq` trait weird. First, all current concrete runqueue types have an empty implementation for this `dequeue` method. So it is questionable whether having this method in the trait is necessary. Second, technically, the `pick_next` method already _dequeues_ the next thread. So the name of the `pick_next` method is misleading. Third, the `update_current` method is also weird. Because the "current" thread is already dequeued.

With that said, I don't have a good alternative design at hand. I just wanted to point out the problem. Maybe you would feel the same.